File: compiler/bench/bellman.fs
\ expected: 19900
\ Bellman-Ford shortest path, 200 nodes, sum distances

200 constant N
400 constant E
1000000 constant INF
create edges E 3 * cells allot
create dist N cells allot
variable nedges

: add-edge ( from to w -- )
  nedges @ 3 * cells edges +
  rot over !
  swap over cell+ !
  cell+ cell+ !
  1 nedges +! ;

: init-graph ( -- )
  0 nedges !
  N 1 do
    i 1- i 1 add-edge
  loop
  N 50 do
    i i 50 - 2 add-edge
  loop ;

: bellman ( start -- )
  N 0 do INF i cells dist + ! loop
  0 swap cells dist + !
  N 1 do
    nedges @ 0 do
      i 3 * cells edges + @
      dup cells dist + @ INF < if
        dup cells dist + @
        i 3 * cells edges + cell+ cell+ @
        +
        i 3 * cells edges + cell+ @
        dup cells dist + @
        2 pick < if
          drop cells dist + !
        else 2drop drop then
      else drop then
    loop
  loop ;

: sum-dist ( -- n )
  0 N 0 do
    i cells dist + @ dup INF < if + else drop then
  loop ;

: main init-graph 0 bellman sum-dist . cr ;

File: compiler/bench/bellman.c
// expected: 19900
#include <stdio.h>
#define N 200
#define E 400
#define INF 1000000

int edges[E][3];  // from, to, weight
long dist[N];
int nedges;

void add_edge(int from, int to, int w) {
    edges[nedges][0] = from;
    edges[nedges][1] = to;
    edges[nedges][2] = w;
    nedges++;
}

void init_graph() {
    nedges = 0;
    for (int i = 1; i < N; i++) {
        add_edge(i - 1, i, 1);
    }
    for (int i = 50; i < N; i++) {
        add_edge(i, i - 50, 2);
    }
}

void bellman(int start) {
    for (int i = 0; i < N; i++) dist[i] = INF;
    dist[start] = 0;
    for (int iter = 1; iter < N; iter++) {
        for (int e = 0; e < nedges; e++) {
            int u = edges[e][0];
            int v = edges[e][1];
            int w = edges[e][2];
            if (dist[u] < INF && dist[u] + w < dist[v]) {
                dist[v] = dist[u] + w;
            }
        }
    }
}

int main() {
    init_graph();
    bellman(0);
    long sum = 0;
    for (int i = 0; i < N; i++) {
        if (dist[i] < INF) sum += dist[i];
    }
    printf("%ld\n", sum);
    return 0;
}
